LmDataset.py

```diff
@@ -1378,10 +1378,21 @@ def init_seq_order(self, epoch=None, seq_list=None):
     else:
       num_seqs = self._get_data_len()
       self._seq_order = self.get_seq_order_for_epoch(
-        epoch=epoch, num_seqs=num_seqs, get_seq_len=lambda i: len(self._get_data(key=self.main_source_data_key, line_nr=i)))
+        epoch=epoch, num_seqs=num_seqs, get_seq_len=lambda i: self.get_estimated_seq_length(i))
```

Comment:
First: Why make the `lambda` here? You could just directly pass `self.get_estimated_seq_length`.

But anyway: This usage is wrong. The way you designed `get_estimated_seq_length` is that it works for seq idx **after** it is already sorted (i.e. given the current seq order). 
However, `get_seq_order_for_epoch` expects to get it globally, i.e. for the whole dataset, what I call "real seq idx" in HDFDataset.

Reply:
See also my comment above. The `get_seq_len` you pass to `get_seq_order_for_epoch` should match the `num_seqs` (which normally is `self.get_total_num_seqs()`, and not `self.num_seqs`), and in general should match the corpus seq idx.

Reply:
(My comment above disappeared, because I applied my suggestion myself. :) So here I copy&paste it again.)

To clarify this `seq_idx` further. I think you even have confused this yourself.
* There is the seq idx related to the current epoch (after you called `init_seq_order`) (this is the same seq idx you would use for `load_seqs` or `get_data`).
* There is a seq idx globally for the dataset (independent of sorting, filtering, etc). In other contexts, I call this "real seq idx", or "corpus seq idx".
This real seq idx might not be well defined for all datasets.
If the function `get_corpus_seq_idx` is defined, it will get epoch seq idx -> corpus seq idx.
If `get_all_tags` is defined, the corpus seq idx is an index in this list.

You should make clear which of these two is expected here.
As I understood your reasoning, you want to have this actually considering the current epoch seq order. So it would use the seq idx related to the current epoch. And not the corpus seq idx.
It is enough if you just write (for `get_estimated_seq_length`) (I already committed this):
```
    :param int seq_idx: for current epoch, not the corpus seq idx
```